### PR TITLE
PCGrad / gradient surgery for multi-objective surface+volume loss (all 4 datasets)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training objective combines surface and volume losses with fixed weighting. When these two objectives have conflicting gradients (e.g. minimizing surface MAE pulls parameters one way, volume MSE another), standard SGD/Adam averages the conflicting directions, causing interference that slows convergence and hurts final performance.

**PCGrad (Gradient Surgery)** (Yu et al., 2020) detects when task gradients conflict (negative cosine similarity) and projects conflicting gradients onto the normal plane of the other task's gradient before summing. This eliminates destructive interference while retaining the full magnitude of non-conflicting components.

**Hypothesis**: Applying PCGrad to the (surface_loss, volume_loss) pair will reduce gradient conflict in datasets where both surface and volume targets are predicted, leading to better surface metrics (the primary target) without sacrificing volume quality. TandemFoil and AirfRANS both predict surface AND volume quantities, making them the primary targets. DrivAerML is surface-only (PCGrad will be a no-op there; run anyway for completeness).

## Implementation

PCGrad is a gradient manipulation step — it wraps the optimizer, not the loss:

```python
# Pseudo-code (simplified):
def pcgrad_step(optimizer, losses):
    # losses: list of per-task loss tensors [L_surface, L_volume]
    grads = []
    for loss in losses:
        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        grads.append([p.grad.clone() for p in model.parameters()])
    
    # For each task i, project away components conflicting with task j
    for i in range(len(grads)):
        for j in range(len(grads)):
            if i == j: continue
            for k, (g_i, g_j) in enumerate(zip(grads[i], grads[j])):
                dot = (g_i * g_j).sum()
                if dot < 0:  # conflicting
                    grads[i][k] = g_i - dot / (g_j.norm() ** 2 + 1e-8) * g_j
    
    # Sum projected gradients and apply
    optimizer.zero_grad()
    for k, p in enumerate(model.parameters()):
        p.grad = sum(g[k] for g in grads)
    optimizer.step()
```

Add flag `--pcgrad` to enable PCGrad. The two "tasks" are the surface loss term and the volume loss term (however they are currently separated in the loss computation).

**If the losses are already summed before backprop** (no separate terms), you will need to refactor the loss computation to keep them separate for the PCGrad step, then sum for the standard step when `--pcgrad` is not set.

**Reference**: "Gradient Surgery for Multi-Task Learning" (Yu et al., NeurIPS 2020). A minimal PyTorch implementation: https://github.com/WeiChengTseng/Pytorch-PCGrad

### TandemFoil
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --pcgrad --epochs 999 --wandb_group pcgrad-surgery
```

### AirfRANS
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --pcgrad --epochs 999 --wandb_group pcgrad-surgery
```

### DrivAerML
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --pcgrad --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb_group pcgrad-surgery
```

### TandemFoil Paper
```bash
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --pcgrad --epochs 999 --wandb_group pcgrad-surgery
```

## Current Baselines (targets to beat)

| Dataset | Metric | Current Best | PR |
|---------|--------|-------------|-----|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 |
| AirfRANS | `val_primary/surface_mse` | **0.000627** | #2902 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 |
| TandemFoil Paper | `val_primary/field_mse` | *no baseline yet* | — |

## Expected Outcome

PCGrad has shown consistent improvements in multi-task learning settings, typically 1-3% improvement on the primary task when objectives genuinely conflict. TF and AF are the most likely beneficiaries. DrivAerML (surface-only) should be neutral, serving as a control.

## W&B Diagnostics to Log

Please log the following to W&B during training (very useful for analysis):
- `train/gradient_conflict_rate`: fraction of parameter gradient pairs with negative cosine similarity each step
- `train/surface_loss` and `train/volume_loss` separately (not just total loss)

## Notes

- Frieren's prior experiments: #2832 (DM WD+gc), #2876 (TF MLP ratio), #2953 (Relative L2 Training Loss, closed as dead end with Wave 3)
- PCGrad has NOT been tried before on this codebase
- Note: PCGrad adds ~2x backprop time per step (two backward passes). This is acceptable — the per-epoch wall time doubles but the hypothesis is that fewer epochs are needed to reach the optimum
- If training becomes too slow with PCGrad's two backward passes, try `--pcgrad-approx` (single backward, split gradient by named parameter groups instead) as a fallback

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
